### PR TITLE
Convert bridge to serve single block tasks

### PR DESCRIPTION
**Pull request by njgheorghita**
### What was wrong?
`backfill` mode was not working well. One major issue was that if a single geth request failed from a set of geth requests, the bridge would discard the entire set of responses and move onto the next batch.

Imo, my strategy of using "batched" geth requests was a bit premature. The immediate focus should be functionality and correctness. Once that is achieved we can turn to optimization.

The new strategy consists of spawning a task for a single block at a time. (fetching is done via geth requests)
- fetch & gossip a single block header -> fetch & gossip body -> fetch & gossip receipt

The old strategy...
- fetch an epochs worth of headers  -> fetch & gossip individual bodies -> fetch & gossip individual receipts
- I also implemented some (in hindsight) not so well thought out batching strategies

Running the bridge with the old strategy yields a torrent of log warnings and errors. To be fair, there are still a significant amount of these with the new strategy. This is why I'd like to simplify the strategy as much as possible, until we see a fairly consistent performance. A major source of these warnings is external (utp-rs), so simplifying our internal strategy makes investigating these a simpler debugging process. 

### How was it fixed?
- operate on serving a constructing & gossiping a single block (header, body, receipts) at a time
- moved `FullHeader` type to `trin-bridge` workspace
- use `buffer_unordered` api to manage concurrent requests rather than custom strategy

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

**Review by KolbyML — COMMENTED** [review]
PR looks good small nit pick about
```
                let header_with_proof = HeaderWithProof {
                    header: full_header.header.clone(),
                    proof: BlockHeaderProof::None(SszNone { value: None }),
                };
```
and
```
                let hwp = Bridge::construct_proof(full_header.header.clone(), epoch_acc).await?;
                HistoryContentValue::BlockHeaderWithProof(hwp)
```
It would be nice to have everything done one way instead of abbreiviating it one place and not another. :+1:

File: trin-bridge/src/bridge.rs [hunk 1]
```diff
@@ -118,118 +114,126 @@ impl Bridge {
...
-            self.gossip_postmerge_headers(gossip_batch).await
+    async fn serve_full_block(
+        height: u64,
+        epoch_acc: Option<Arc<EpochAccumulator>>,
+        gossip_stats: Arc<Mutex<GossipStats>>,
+        portal_clients: Vec<HttpClient>,
+    ) -> anyhow::Result<()> {
+        debug!("Serving block: {height}");
+        let mut full_header = Bridge::get_header(height).await?;
+        if full_header.header.number <= MERGE_BLOCK_NUMBER {
```
Comment (ogenev) [inline trin-bridge/src/bridge.rs:168]: I think those kinds of comparisons reveal that we may need multiple `FullHeader`  versions represented as an enum.  I'm not saying that this is in the scope of this PR but something that we may think about or create an issue.

File: trin-bridge/src/bridge.rs [hunk 2]
```diff
@@ -118,118 +114,126 @@ impl Bridge {
...
+                if header_record.block_hash != full_header.header.hash() {
+                    bail!(
+                        "Header hash doesn't match record in local accumulator: {:?} - {:?}",
+                        full_header.header.hash(),
+                        header_record.block_hash
+                    );
+                }
+                // Construct HeaderWithProof
+                let hwp = Bridge::construct_proof(full_header.header.clone(), epoch_acc).await?;
+                HistoryContentValue::BlockHeaderWithProof(hwp)
```
[the remark this answers is not in the document]
Comment (ogenev) [reply trin-bridge/src/bridge.rs:209]: I agree that `header_with_proof` is more readable.

**Review by ogenev — APPROVED** [review]
:shipit:

File: trin-bridge/src/full_header.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            headers: results.map_err(serde::de::Error::custom)?,
+        })
+    }
+}
+
+/// Datatype for use in bridge functionality. The purpose is a single datatype that can be created
+/// from a header and contains all the information necessary to build history network content
+/// values for BlockBodies (txs, uncles) and Receipts (tx_hashes) through subsequnt jsonrpc
+/// requests.
+#[derive(Debug, Clone, PartialEq, Eq)]
```
[the remark this answers is not in the document]
Comment (njgheorghita) [reply trin-bridge/src/full_header.rs:40]: Using `Deserialize` trait is always preferable. The primary reason I went with custom implementations here is to deal with json responses from geth, primarily handling the `"result"` field. Agree there could be a better solution here, but it's not a priority for me right now. ... tracking in #754

File: trin-bridge/src/bridge.rs (hunk 1 above)
Comment (njgheorghita) [reply trin-bridge/src/bridge.rs:173]: Agree that your instinct is correct here. However, I'm not convinced that the enum overhead validates implementation for a fairly simple check that we do twice. But, I'd be curious to see what it looks like if somebody wants to give it a go. ... tracking in #754

**Pull request merged**